src/lib/logStorage.ts

```
@@ -0,0 +1,162 @@
+import { openDB, type IDBPDatabase } from 'idb';
+
+const DB_NAME = 'dgen-logs';
+const DB_VERSION = 1;
+const STORE_NAME = 'logs';
+
+const MAX_LOGS = 5000;
+const CLEANUP_INTERVAL = 100;
+const CLEANUP_BUFFER = 100;
+
+let dbPromise: Promise<IDBPDatabase> | null = null;
+let writesSinceCleanup = 0;
+
+const WRITE_BATCH_SIZE = 20;
+const WRITE_FLUSH_INTERVAL_MS = 500;
+
+let pendingLines: string[] = [];
+let flushTimeout: number | null = null;
+let isFlushing = false;
+
+async function getDB(): Promise<IDBPDatabase> {
+  if (!dbPromise) {
+    dbPromise = openDB(DB_NAME, DB_VERSION, {
+      upgrade(db, oldVersion) {
+        if (oldVersion < 1) {
+          if (db.objectStoreNames.contains(STORE_NAME)) {
+            db.deleteObjectStore(STORE_NAME);
+          }
+          db.createObjectStore(STORE_NAME, { autoIncrement: true });
+        }
+        // Future migrations: if (oldVersion < 2) { ... }
+      },
+    }).catch(err => {
+      dbPromise = null;
+      console.error('[logStorage] Failed to open database:', err);
+      throw err;
+    });
+  }
+  return dbPromise;
+}
+
+function isBrowser(): boolean {
+  return typeof window !== 'undefined' && typeof indexedDB !== 'undefined';
+}
+
+async function flushPendingLines(): Promise<void> {
+  if (!isBrowser()) return;
+  if (isFlushing) return;
+  if (pendingLines.length === 0) return;
+
+  isFlushing = true;
+
+  const batch = pendingLines;
+  pendingLines = [];
```

If `getDB()` or the transaction fails, the batch is lost forever. The error is logged but logs are silently dropped.

On failure, prepend the batch back to `pendingLines`:
```
} catch (err) {
    console.error('[logStorage] Failed to flush logs batch:', err);
    pendingLines = [...batch, ...pendingLines]; // restore failed batch
}